### PR TITLE
[6.x] Allow Storage::put to accept a Psr StreamInterface

### PR DESCRIPTION
**Pull request by Gman98ish**
Addresses: https://github.com/laravel/ideas/issues/1252

This PR will allow the Storage::put method to accept a Psr StreamInterface. This will mostly be useful when dealing with guzzle, as you will no longer have to manually call detach and pass a resource to Storage::put.

This will not break existing features as it's broadening what the method accepts.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
         "opis/closure": "^3.1",
         "psr/container": "^1.0",
         "psr/simple-cache": "^1.0",
+        "psr/http-message": "^1.0",
```
Comment (GrahamCampbell) [inline composer.json]: Please add this as a suggested dependency, and also update the `src/Illuminate/Filesystem/composer.json` file.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (Gman98ish) [reply composer.json]: Thanks for the feedback. I've changed it to a suggested dependency in the root composer.json and added it as a suggested dependency in the filesystem composer.json

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
Please order the suggested dependencies alphabetically.

**Comment by netpok** [comment]
I am not sure about this, because it calls detach behind the scenes, I wouldn't anticipate that a filesystem write command leaves my stream in an [unusable state](https://github.com/php-fig/http-message/blob/master/src/StreamInterface.php#L40).

Or it should be very well documented.

**Comment by taylorotwell** [comment]
@Gman98ish any comments on @netpok's comment?

**Comment by Gman98ish** [comment]
@netpok That's a good point, it could be pretty confusing/irritating. I guess the alternative would be to avoid using detach and instead read the contents of the stream into memory, then use the regular put method?

Obvious problem with that is that you have the whole contents of the stream in memory, which sorta defeats the point of using a stream. I also looked into chunking it without the use of putStream, but it doesn't look like you can consistently append to a file without relying on the config of the flysystem adapter.

I'm happy to drop this if the overall behaviour is too confusing. Or at the very least, come back to it when I have a better solution

**Comment by taylorotwell** [comment]
I think I will merge it for now. We were already calling putStream for other types of streams.

**Pull request merged**